### PR TITLE
Fix Android keyboard in hot mode

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Rework how application restart works in hot mode so that the PlatformView can delegate to the native OS thread so that services are properly wired up.
- [x] Fix keyboard on Android.

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@chinmaygarde @abarth @jason-simmons

File: sky/shell/platform/android/io/flutter/view/FlutterView.java [hunk 1]
```diff
@@ -472,13 +473,50 @@ public void runFromBundle(String bundlePath, String snapshotPath) {
...
+            this.main = main;
+            this.packages = packages;
+            this.assetsDirectory = assetsDirectory;
+            this.flutterView = flutterView;
+        }
+
+        private String main;
+        private String packages;
+        private String assetsDirectory;
+        private FlutterView flutterView;
```
Comment (jason-simmons) [inline sky/shell/platform/android/io/flutter/view/FlutterView.java]: The RunFromSourceRunnable inner class will have an implicit reference to the outer FlutterView.

You can make the arguments to runFromSource final and reduce RunFromSourceRunnable to an anonymous inner class within runFromSource.

Comment (johnmccutchan) [inline sky/shell/platform/android/io/flutter/view/FlutterView.java]: Done.

File: sky/shell/platform/android/io/flutter/view/FlutterView.java [hunk 2]
```diff
@@ -472,13 +473,50 @@ public void runFromBundle(String bundlePath, String snapshotPath) {
...
+
+    public void runFromSource(String main,
+                              String packages,
+                              String assetsDirectory) {
+        RunFromSourceRunnable runnable =
+            new RunFromSourceRunnable(main, packages, assetsDirectory, this);
+        post(runnable);
+        try {
+            synchronized (runnable) {
+                runnable.wait();
```
Comment (jason-simmons) [inline sky/shell/platform/android/io/flutter/view/FlutterView.java]: There is a potential race here if the task calls notify() before this thread enters the wait().  Move the post() call into the synchronized block, or consider android.os.ConditionVariable or java.util.concurrent.Semaphore

File: sky/shell/platform/android/platform_view_android.cc [hunk 3]
```diff
@@ -343,10 +344,16 @@ class AndroidGLContext {
...
 
-static jlong Attach(JNIEnv* env, jclass clazz, jint skyEngineHandle) {
+static jlong Attach(JNIEnv* env,
+                    jclass clazz,
+                    jint skyEngineHandle,
+                    jobject flutterView) {
   PlatformViewAndroid* view = new PlatformViewAndroid();
   view->ConnectToEngine(mojo::InterfaceRequest<SkyEngine>(
       mojo::ScopedMessagePipeHandle(mojo::MessagePipeHandle(skyEngineHandle))));
+  // Remember the JNIEnv and create a weak reference to the flutterView Java
```
Comment (jason-simmons) [inline sky/shell/platform/android/platform_view_android.cc]: "Remember the JNIEnv" is probably obsolete?  RunFromSource will attach the VM to get a JNIEnv

File: sky/shell/platform/android/io/flutter/view/FlutterView.java (hunk 2 above)
Comment (johnmccutchan) [inline sky/shell/platform/android/io/flutter/view/FlutterView.java]: Done.

File: sky/shell/platform/android/platform_view_android.cc (hunk 3 above)
Comment (johnmccutchan) [inline sky/shell/platform/android/platform_view_android.cc]: Yes, comment was written while I was learning how to use JNI. Comment fixed.

File: sky/shell/platform/android/platform_view_android.cc [hunk 4]
```diff
@@ -435,5 +444,48 @@ void PlatformViewAndroid::Resize(const SkISize& size) {
...
+  jstring java_assets_directory = env->NewStringUTF(assets_directory.c_str());
+  FTL_CHECK(java_assets_directory);
+  env->CallVoidMethod(local_flutter_view,
+                      run_from_source_method_id,
+                      java_main,
+                      java_packages,
+                      java_assets_directory);
+
+  // Delete the local reference.
+  env->DeleteLocalRef(local_flutter_view);
```
Comment (jason-simmons) [inline sky/shell/platform/android/platform_view_android.cc]: Local references (the local_flutter_view, plus flutter_view_class and the strings) will be deleted when the thread detaches.

I'd suggest consistently deleting all of the references (possibly with ScopedJavaLocalRef or ScopedJavaLocalThread) or explicitly noting that the detach will delete the local references.

File: sky/shell/platform/android/platform_view_android.cc [hunk 5]
```diff
@@ -343,10 +344,16 @@ class AndroidGLContext {
...
+static jlong Attach(JNIEnv* env,
+                    jclass clazz,
+                    jint skyEngineHandle,
+                    jobject flutterView) {
   PlatformViewAndroid* view = new PlatformViewAndroid();
   view->ConnectToEngine(mojo::InterfaceRequest<SkyEngine>(
       mojo::ScopedMessagePipeHandle(mojo::MessagePipeHandle(skyEngineHandle))));
+  // Remember the JNIEnv and create a weak reference to the flutterView Java
+  // object so that we can make calls into it later.
+  view->set_flutter_view(env->NewWeakGlobalRef(flutterView));
```
Comment (jason-simmons) [inline sky/shell/platform/android/platform_view_android.cc]: When is this deleted?

The JavaObjectWeakGlobalRef wrapper in base can help with this

Comment (johnmccutchan) [inline sky/shell/platform/android/platform_view_android.cc]: Done.

**Comment by johnmccutchan** [comment]
PTAL

File: sky/shell/platform/android/platform_view_android.h [hunk 6]
```diff
@@ -52,9 +53,18 @@ class PlatformViewAndroid : public PlatformView {
   // sky::shell::PlatformView override
   virtual void Resize(const SkISize& size);
 
+  virtual void RunFromSource(const std::string& main,
+                             const std::string& packages,
+                             const std::string& assets_directory);
+
+  void set_flutter_view(JavaObjectWeakGlobalRef flutter_view) {
```
Comment (jason-simmons) [inline sky/shell/platform/android/platform_view_android.h]: can pass a const JavaObjectWeakGlobalRef& here

**Comment by jason-simmons** [comment]
LGTM

File: sky/shell/platform/android/platform_view_android.h (hunk 6 above)
Comment (johnmccutchan) [inline sky/shell/platform/android/platform_view_android.h]: Done. Thanks for all the Java / Android / JNI linting -- I appreciate it.

**Pull request merged**